docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,137 @@
+finding rbd image based on PVC name and namespace name in the remote cluster)
+
+The value for the mirroring will be  either one of these "disabled" (default),
+"journal", or "snapshot"
+
+* "disabled" - (default) no need to mirror the rbd image
+* "journal"  - journal based mirroring. The mounter should be rnd-nbd as the
+  krbd doesnot support it yet.
+* "snapshot" - snapshot based mirroring. It should work with existing krbd
+  (need ceph octopus+)
+
+**Note**: Older PVC or Snapshots cannot be mirrored
+
+More info on rbd mirroring at [1].
+
+### Pass PVC name and namespace name in CreateVolume
+
+If the app and the PVC created in the local cluster as part of recovery the app
+and the PVC will be recreated in the remote cluster, the data present in the
+PVC in remote  cluster should be the same in the local cluster.
+
+As the PVC name and the namespace name should be sent in volume create request,
+cephcsi will check that the rbd image is already present cluster, we cannot
+check the omap data first as omap data maintained by ceph-csi is not mirrored
+to the remote cluster. This is the challenge ceph-csi need to handle)If the
+image is not present(means its a local cluster) we need to generate the
+required omap data and store its information and create the rbd image, the CSI
+will need to enable mirroring on the newly created RBD image via `rbd mirror
+image enable image-spec`.
+
+If it is a remote cluster the image will be already created and mirrored by the
+local cluster, we need to check for the image first(if it is present we need to
+generate the omap data and return success to the create volume (as its a remote
+cluster we don't need to enable mirroring here.
+
+Incase of journal based mirroring the image need to be craeted with
+exclusive-lock and journaling image features. Additionally, This will also
+imply that NBD will need to be used instead of krbd.
+
+### Promotion/Demotion of the rbd image
+
+During the **NodePublish request** from the CSI side, ceph-csi will
+promote the image and use it. and during **NodeUnpublish request** the ceph-csi
+will  demote the image and use it.
+
+as the NodePublish is per node so cephcsi  need to handle if the image is
+already promoted.
+
+But during **NodeUnpublish request** the ceph-csi should make sure that the rbd
+image is not attached on any node  before demoting the image as it can cause
+data lose. May be cephcsi can make use of volumeattachment kubernetes object to
+check the pvc is attached to any nodes or not.
+
+If its a ROX or RWX PVC we need to handle image already mirrored case.
+
+#### Incase of dead primary/remote cluster
+
+In case the ceph cluster is dead ceph-csi cannot demote the image to over come
+this issue we will provide a new flag (configuration) in cephcsi which will be
+set by the high level operator who is responsible for data migration will set
+the flag to `true` so that CSI can forcefull promote the image.
```

Comment:
I understand this was the discussion we had. However we have to make sure the higher level operator should make sure there is a complete fence of the access from the other end to avoid data corruption and other issues due to this force enablement.